Core/GameEngine/Source/Common/System/LocalFile.cpp

```diff
@@ -167,61 +167,75 @@ Bool LocalFile::open( const Char *filename, Int access )
 	}
 
 	/* here we translate WSYS file access to the std C equivalent */
-#ifdef USE_BUFFERED_IO
-	char mode[32];
-	char* m = mode;
-
-	if (m_access & APPEND)
-	{
-		DEBUG_CRASH(("not yet supported by buffered mode"));
-	}
-
-	if (m_access & TRUNCATE)
-	{
-		DEBUG_CRASH(("not yet supported by buffered mode"));
-	}
-
-	if((m_access & READWRITE ) == READWRITE )
+#if USE_BUFFERED_IO
+
+	// r    open for reading (The file must exist)
+	// w    open for writing (creates file if it doesn't exist). Deletes content and overwrites the file.
+	// a    open for appending (creates file if it doesn't exist).
+	// r+   open for reading and writing (The file must exist).
+	// w+   open for reading and writing.
+	//      If file exists deletes content and overwrites the file, otherwise creates an empty new file.
+	// a+   open for reading and writing (append if file exists).
+
+	const Bool write     = (m_access & WRITE) != 0;
+	const Bool readwrite = (m_access & READWRITE) == READWRITE;
+	const Bool append    = (m_access & APPEND) != 0;
+	const Bool create    = (m_access & CREATE) != 0;
+	const Bool truncate  = (m_access & TRUNCATE) != 0;
+	const Bool text      = (m_access & TEXT) != 0;
+	const Bool binary    = (m_access & BINARY) != 0;
+
+	const Char *mode = NULL;
+
+	// Mode string selection (mimics _open flag combinations)
+	// TEXT is implicit for fopen if 'b' is not present
+	// READ is implicit here if not READWRITE or WRITE
+	if (readwrite)
 	{
-		if (m_access & CREATE)
-		{
-			*m++ = 'w';
-			*m++ = '+';
-		}
+		if (append)
+			mode = binary ? "a+b" : "a+";
+		else if (truncate || create)
+			mode = binary ? "w+b" : "w+";
 		else
-		{
-			*m++ = 'r';
-			*m++ = '+';
-		}
+			mode = binary ? "r+b" : "r+";
 	}
-	else if(m_access & WRITE)
+	else if (write)
 	{
-		*m++ = 'w';
+		if (append)
+			mode = binary ? "ab" : "a";
+		else
+			mode = binary ? "wb" : "w";
 	}
-	else
+	else // implicitly read-only
 	{
-		*m++ = 'r';
-		DEBUG_ASSERTCRASH(!(m_access & TRUNCATE), ("cannot truncate with read-only"));
+		mode = binary ? "rb" : "r";
 	}
 
-	if (m_access & TEXT)
+	m_file = fopen(filename, mode);
+	if (m_file == NULL)
 	{
-		*m++ = 't';
+		goto error;
 	}
-	if (m_access & BINARY)
+
 	{
-		*m++ = 'b';
-	}
+		Int result = 0;
 
-	*m++ = 0;
+		if (bufferSize == 0)
+		{
+			result = setvbuf(m_file, NULL, _IONBF, 0); // Uses no buffering.
+		}
+		else
+		{
+			const Int bufferMode = text
+				? _IOLBF // Uses line buffering
```

Comment:
We may only want line buffering on `text && write` ?

Reply:
After some thinking I concluded that it is better separated from TEXT and BINARY, so I added new modes LINEBUF and FULLBUF. This way the caller has full control over the behavior.